### PR TITLE
[ios][web-browser] Rewrite the module to Sweet API

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

Fixes ENG-3067 which is a subtask of the `expo-web-browser` audit

# How

It turned out the existing code in Objective-C is pretty messy and I just found it easier to rewrite it to Swift from scratch instead of mirroring the functionality.
There are some things that I intentionally changed/removed:
- Saving and restoring the status bar state (looks like we don't need it, I'll verify it later and bring it back if needed)
- Replaced deprecated `SFAuthenticationSession` with new `ASWebAuthenticationSession`
- When dismissing the browser, I dismiss the browser controller instead of the "current" one
- The hack with additional modal navigation controller doesn't seem to make the difference so I removed that for now

# Test Plan

NCL examples seem to work as expected. I tested it against some more examples, but I'll add them separately.
SwiftLint has no complaints.

### CODE REVIEW

File: packages/expo-web-browser/ios/WebBrowserSession.swift [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    viewController.dismissButtonStyle = options.dismissButtonStyle.toSafariDismissButtonStyle()
+    viewController.preferredBarTintColor = options.toolbarColor
+    viewController.preferredControlTintColor = options.controlsColor
+
+    super.init()
+    viewController.delegate = self
+
+    // By setting the modal presentation style to OverFullScreen, we disable the "Swipe to dismiss"
+    // gesture that is causing a bug where sometimes `safariViewControllerDidFinish` is not called.
+    // There are bugs filed already about it on OpenRadar.
```
Comment (bbarthec) [inline packages/expo-web-browser/ios/WebBrowserSession.swift:29]: Maybe add a references to the bugs here? 🤔

File: packages/expo-web-browser/ios/WebBrowserSession.swift [hunk 2]
```diff
@@ -0,0 +1,62 @@
+// Copyright 2022-present 650 Industries. All rights reserved.
+
+import SafariServices
+import ExpoModulesCore
+
+internal class WebBrowserSession: NSObject, SFSafariViewControllerDelegate {
+  let viewController: SFSafariViewController
+  var modalNavigationController: UINavigationController?
```
Comment (bbarthec) [inline packages/expo-web-browser/ios/WebBrowserSession.swift:8]: This seems unneeded as you're only referencing in a single place in the code by `modalNavigationController = nil`
```suggestion
```

File: packages/expo-web-browser/ios/WebBrowserSession.swift [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  func safariViewControllerDidFinish(_ controller: SFSafariViewController) {
+    finish(type: "cancel")
+  }
+
+  // MARK: - Private
+
+  private func finish(type: String) {
+    promise?.resolve(["type": type])
+    promise = nil
+    modalNavigationController = nil
```
Comment (bbarthec) [inline packages/expo-web-browser/ios/WebBrowserSession.swift:60]: I don't see you're clearing the `viewController` anywhere 👁️ 
```suggestion
    viewController = nil
```

File: packages/expo-web-browser/ios/WebBrowserModule.swift [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    name("ExpoWebBrowser")
+
+    function("openBrowserAsync") { (url: URL, options: WebBrowserOptions, promise: Promise) throws in
+      guard self.currentWebBrowserSession?.isOpen != true else {
+        throw WebBrowserAlreadyOpenException()
+      }
+      self.currentWebBrowserSession = WebBrowserSession(url: url, options: options)
+      self.currentWebBrowserSession?.open(promise)
+    }
+    .runOnQueue(.main)
```
Comment (bbarthec) [inline packages/expo-web-browser/ios/WebBrowserModule.swift:21]: Isn't that a default behaviour? 🤔

**Review by bbarthec — COMMENTED** [review]
Awesome work 👏 
Really easy to read and follow the logic 😁
Left some remarks 👇

File: packages/expo-web-browser/ios/WebBrowserSession.swift (hunk 1 above)
Comment (tsapeta) [reply packages/expo-web-browser/ios/WebBrowserSession.swift:29]: If only I have the references 😂 I just kept this comment and logic with the hope I'll address it later. There were several things that I was not sure about and would rather like to test/fix them one by one.

File: packages/expo-web-browser/ios/WebBrowserModule.swift (hunk 4 above)
Comment (tsapeta) [reply packages/expo-web-browser/ios/WebBrowserModule.swift:21]: I thought it is, but I was wrong. We use `DispatchQueue.global(qos: .default)` as a default queue, which doesn't have to run on the main thread. I think we'll need to change it in this cycle + add kind of `defaultQueue` component to set the default queue for all module's functions

**Review by bbarthec — APPROVED** [review]
Looking good 👍 Let's ship it 🚢

**Comment by gregdburns** [comment]
@tsapeta  / @byCedric - would it be possible to expose [prefersEphemeralWebBrowserSession](https://developer.apple.com/documentation/authenticationservices/aswebauthenticationsession/3237231-prefersephemeralwebbrowsersessio) as a part of this PR?  While this would solve it for iOS only, it would allow people to run the session in a state where cookies aren't shared, which seems to be a common challenge: https://github.com/expo/expo/issues/6756, etc.

**Comment by bbarthec** [comment]
@gregdburns, it would be best to add this feature you're asking about in the separate PR. You can either create an issue requesting for that or even propose the PR on your own! 🤩

**Pull request merged**